File: 1/2020.10.05/127.Word Ladder/java/solution.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if (endWord == null || wordList == null || !wordList.contains(endWord)) {
            return 0;
        }
        Map<String, List<String>> allComboDict = new HashMap<>();
        int n = beginWord.length();
        for (String word : wordList) {
            for (int i = 0; i < n; i++) {
                String key = word.substring(0, i) + "*" + word.substring(i + 1, word.length());
                if (!allComboDict.containsKey(key)) {
                    allComboDict.put(key, new ArrayList<>());
                }
                allComboDict.get(key).add(word);
            }
        }
        Deque<Pair<String, Integer>> qBegin = new LinkedList<>();
        Deque<Pair<String, Integer>> qEnd = new LinkedList<>();
        qBegin.offerLast(new Pair<>(beginWord, 1));
        qEnd.offerLast(new Pair<>(endWord, 1));
        Map<String, Integer> visitedBegin = new HashMap<>();
        Map<String, Integer> visitedEnd = new HashMap<>();
        visitedBegin.put(beginWord, 1);
        visitedEnd.put(endWord, 1);
        while (!qBegin.isEmpty() && !qEnd.isEmpty()) {
            int res = visitWord(qBegin, visitedBegin, visitedEnd, allComboDict);
            if (res != -1) {
                return res;
            }
            res = visitWord(qEnd, visitedEnd, visitedBegin, allComboDict);
            if (res != -1) {
                return res;
            }
        }
        return 0;
    }

    private int visitWord(Deque<Pair<String, Integer>> q, Map<String, Integer> visited, Map<String, Integer> otherVisited, Map<String, List<String>> allComboDict) {
        Pair<String, Integer> pair = q.pollFirst();
        for (int i = 0; i < pair.getKey().length(); i++) {
            String intermediateWord = pair.getKey().substring(0, i) + "*" + pair.getKey().substring(i + 1, pair.getKey().length());
            if (allComboDict.containsKey(intermediateWord)) {
                for (String word : allComboDict.get(intermediateWord)) {
                    if (otherVisited.containsKey(word)) {
                        return otherVisited.get(word) + pair.getValue();
                    }
                    if (!visited.containsKey(word)) {
                        q.offerLast(new Pair<>(word, pair.getValue() + 1));
                        visited.put(word, pair.getValue() + 1);
                    }
                }
            }
        }
        return -1;
    }
}

File: 1/2020.10.05/127.Word Ladder/python/solution.py
from collections import defaultdict
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        # BFS
        if not endWord or not wordList or endWord not in wordList:
            return 0
        n = len(beginWord)
        all_combo_dict = defaultdict(list)
        for word in wordList:
            for i in range(n):
                all_combo_dict[word[:i] + '*' + word[i+1:]].append(word)
        
        q = [(beginWord, 1)]
        visited = {beginWord: True}
        while q:
            current_word, length = q.pop(0)
            for i in range(n):
                intermediate_word = current_word[:i] + '*' + current_word[i+1:]
                for word in all_combo_dict[intermediate_word]:
                    if word == endWord:
                        return length + 1
                    if word not in visited:
                        q.append((word, length + 1))
                        visited[word] = True
                all_combo_dict[intermediate_word] = []
        return 0

    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        # double DFS
        def visit_word(q, visited, other_visited):
            current_word, length = q.pop(0)
            for i in range(n):
                intermediate_word = current_word[:i] + '*' + current_word[i+1:]
                for word in all_combo_dict[intermediate_word]:
                    if word in other_visited:
                        return length + other_visited[word]
                    if word not in visited:
                        q.append((word, length+1))
                        visited[word] = length+1
            return None

        if not endWord or not wordList or endWord not in wordList:
            return 0
        n = len(beginWord)
        all_combo_dict = defaultdict(list)
        for word in wordList:
            for i in range(n):
                all_combo_dict[word[:i] + '*' + word[i+1:]].append(word)

        q_begin = [(beginWord, 1)]
        q_end = [(endWord, 1)]
        visited_begin = {beginWord: 1}
        visited_end = {endWord: 1}

        while q_begin and q_end:
            res = visit_word(q_begin, visited_begin, visited_end)
            if res:
                return res
            res = visit_word(q_end, visited_end, visited_begin)
            if res:
                return res
        return 0
